main.cpp

```diff
@@ -21,41 +24,72 @@ cstring remove_extension(cstring filename) {
     return filename.substr(0, idx);
 }
 
+const IR::P4Program *prune_statements(const IR::P4Program *program,
+                                      P4PRUNER::PrunerOptions options,
+                                      int required_exit_code) {
+    cstring stripped_name = remove_extension(options.file);
+    stripped_name += "_stripped.p4";
+    for (int j = 0; j < 20; j++) {
+        P4PRUNER::Pruner *pruner = new P4PRUNER::Pruner();
+
+        auto temp = program;
+        temp = temp->apply(*pruner);
+        auto temp_f = new std::ofstream(stripped_name);
+
+        P4::ToP4 *temp_p4 = new P4::ToP4(temp_f, false);
+        temp->apply(*temp_p4);
+        cstring new_command = "python3 ";
+        new_command += realpath(options.validator_script, NULL);
+        new_command += " -i ";
+        new_command += stripped_name;
+        int exit_code = system(new_command.c_str());
+
+        if (exit_code == required_exit_code) {
+            program = temp;
+        }
+    }
+    return program;
+}
+
 int main(int argc, char *const argv[]) {
     AutoCompileContext autoP4toZ3Context(new P4PRUNER::P4PrunerContext);
     auto &options = P4PRUNER::P4PrunerContext::get().options();
     options.langVersion = CompilerOptions::FrontendVersion::P4_16;
+    const IR::P4Program *program;
 
     if (options.process(argc, argv) != nullptr) {
         options.setInputFile();
     }
     if (::errorCount() > 0)
         return 1;
 
-    if (options.file == nullptr) {
+    if (options.file == nullptr || options.validator_script == nullptr) {
         options.usage();
         return 1;
     }
 
-    const IR::P4Program *program = nullptr;
+    INFO("\nChecking " << options.file);
+    // Assemble the validator command
```

Comment:
We probably should add comments to each program block explaining its purpose. But avoid overcommenting.